src/swarm/neo/util/StructSerializer.d

```diff
@@ -0,0 +1,349 @@
+/*******************************************************************************
+
+    Simple contiguous memory serializer that accepts reference types.
+
+    Useful for serializing request contexts, which often contain things like
+    delegates, unions, and pointers. (Ocean's contiguous serializer rejects such
+    types, for good reason.)
+
+    Copyright:
+        Copyright (c) 2016-2017 sociomantic labs GmbH. All rights reserved
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
+module swarm.neo.util.StructSerializer;
+
+import ocean.transition;
+import ocean.core.Traits;
+
+/*******************************************************************************
+
+    Serializes `s` into the provided buffer, ensuring that the content of all
+    dynamic array fields are contained fully within the buffer.
+
+    Params:
+        S = type to serialize
+        s = thing to serialize
+        buf = buffer to serialize into
+
+*******************************************************************************/
+
+public void serialize ( S ) ( S s, ref void[] buf )
+{
+    static assert(canSerialize!(S)());
+
+    // Do a simple, binary copy of the struct.
+    buf.length = S.sizeof;
+    enableStomping(buf);
+    buf[] = (cast(void*)&s)[0..S.sizeof];
+
+    // Handle any dynamic array fields. (Note that it's safe to cast away const
+    // at this point as we're referring to the copy of s inside buf, no longer
+    // to s itself.)
+    auto serialized_struct = cast(Unqual!(S)*)buf.ptr;
+    seriaizeStructDynamicArrays(*serialized_struct, buf);
+}
+
+/*******************************************************************************
+
+    Interprets the data in the provided buffer (which must have been created by
+    a call to `serialize`, above) as a struct of the specific type.
+
+    Params:
+        S = type to deserialize
+        serialized = buffer containing the serialized thing
+
+    Returns:
+        pointer to the buffer interpreted as a pointer to an instance of the
+        serialized type
+
+*******************************************************************************/
+
+public S* deserialize ( S ) ( Const!(void)[] serialized )
```

Comment:
This casts the `const` away from `serialized`. It looks like a case for `inout`,
```d
Inout!(S)* deserialize ( S ) ( Inout!(void)[] serialized )
```
so that the qualifier of the return type follows that of the argument.

Reply:
Mm I don't see why that would be useful...?

1. `Const!(void)[] serialized` states that we never modify the input buffer.
2. `S*` states that we return a pointer to some type. (Note that the user can do `auto s = deserialize!(Const!(Whatever))(buffer);` if they want to.)